keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+- exactly one of the fields `indexes` and `count` is specified
+  for a requirement.
+- whether the specified indexes in the `indexes` and
+  the number of indexes in the `count` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `indexes` is under 64Ki.
+- exactly one of the enum `continue` or `terminate` is specified
+in the `whenTargetAchieved`.
+
+In addition, when the `successPolicy` isn't nil, we validate the `onSucceededPods` isn't nil.
+
+### Evaluation
+We evaluate the SuccessPolicy every in updating pod conditions.
+And then if the job meets the SuccessPolicy, we add `Completed` condition
+instead of `Failed` condition to `.status.conditions` even if some indexes fail.
+
+If a set of indexes are specified in the `indexes` field,
+we evaluate `.status.completedIndexes` to see if a set of indexes is there.
+
+If the number of indexes is specified in the `count`, 
+we evaluate `.status.succeeded` to see if the value is `count` or more.
+
+Also, the job controller do actions against running pods based on the value of `onSucceededPods.whenTargetAchieved`:
+
+- `continue`: Nothing. Running pods continues to run.
+- `terminate`: Not finished pods would be terminated by the job controller. Not finished pods means running or pending pods.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+and backoffLimit in the following situations: 
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+
+So, the successPolicy will be evaluated after the pod failure policy and backoffLimit to avoid the above conflicts.
+Therefore, the pod failure policy and backoffLimit will be respected rather than the successPolicy.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test cases:
+  - tests for Defaulting and Validating
+  - tests for rounding up absolute numbers in `.spec.successPolicy.onSucceededPods.count`.
+  - verify whether a job has condition=completed if the job meets to successPolicy
+  and some indexes fail.
+
+- `k8s.io/kubernetes/pkg/controller/job`: `6 June 2023` - `90.5%`
+- `k8s.io/kubernetes/pkg/apis/batch/validation`: `6 June 2023` - `98.5%`
+
+##### Integration tests
+
+- Test scenarios:
+  - enabling, disabling and re-enabling of the `JobSuccessPolicy` feature gate
+  - handling of successPolicy when all indexes succeeded
+  - handling of the `.spec.successPolicy.onSucceededPods.indexes` when some indexes fail
```

Comment:
```suggestion
  - handling of the `.spec.successPolicy.onSucceededPods.indexes` when some indexes remain pending
```
I don't think there is a concept of index failing prior to `backoffLimitPerIndex` feature

Reply:
Done.